File: services/schedule_planner.py
from services.time_utils import format_time, minutes_to_time, time_to_minutes


def compute_schedule(stops: list, segment_minutes: list, start_time: str):
    """Berechnet ETA/ETD entlang der Stoppreihenfolge.

    Start ist die Tour-Startzeit. Für jeden Stopp gilt:
    Ankunft = vorige Abfahrt + Fahrzeit. Liegt die Ankunft vor dem Zeitfenster,
    wartet der Plan bis zum Zeitfenster-Start. Liegt sie nach dem Zeitfenster-Ende,
    wird ein Konflikt markiert. Abfahrt = effektive Ankunft + Aufenthaltszeit.

    `segment_minutes` enthält die Fahrzeit vom Depot zum ersten Stopp, dann zwischen
    Stopps und optional zuletzt die Rückfahrt zum Depot. Für ETA/ETD der Stopps werden
    nur die ersten `len(stops)` Segmente benötigt.
    """

    result_stops = []
    segment_details = []
    current_minutes = time_to_minutes(start_time)
    if current_minutes is None:
        current_minutes = 8 * 60

    total_travel_minutes = 0
    total_service_minutes = 0
    total_wait_minutes = 0
    schedule_blocked = False

    for index, stop in enumerate(stops):
        segment_value = segment_minutes[index] if index < len(segment_minutes) else None
        if segment_value is not None:
            total_travel_minutes += segment_value

        arrival_minutes = None if schedule_blocked or segment_value is None else current_minutes + segment_value
        arrival_text = format_time(minutes_to_time(arrival_minutes)) if arrival_minutes is not None else ""

        window_start = time_to_minutes(stop.get("time_window_start"))
        window_end = time_to_minutes(stop.get("time_window_end"))
        wait_minutes = 0
        conflict_text = ""
        conflict = False

        effective_arrival = arrival_minutes
        if effective_arrival is not None and window_start is not None and effective_arrival < window_start:
            wait_minutes = window_start - effective_arrival
            effective_arrival = window_start
            total_wait_minutes += wait_minutes

        if effective_arrival is not None and window_end is not None and effective_arrival > window_end:
            conflict = True
            conflict_text = f"Ankunft {format_time(minutes_to_time(effective_arrival))} > Fenster Ende {stop.get('time_window_end')}"

        service_minutes = int(stop.get("service_minutes") or 0)
        departure_minutes = None if effective_arrival is None else effective_arrival + service_minutes
        if departure_minutes is not None:
            total_service_minutes += service_minutes

        if effective_arrival is None:
            schedule_blocked = True
        else:
            current_minutes = departure_minutes

        updated = dict(stop)
        updated["planned_arrival"] = arrival_text
        updated["planned_departure"] = format_time(minutes_to_time(departure_minutes)) if departure_minutes is not None else ""
        updated["schedule_conflict"] = conflict
        updated["schedule_conflict_text"] = conflict_text
        updated["wait_minutes"] = wait_minutes
        result_stops.append(updated)

        segment_details.append(
            {
                "index": index,
                "travel_minutes": segment_value,
                "arrival": updated["planned_arrival"],
                "departure": updated["planned_departure"],
                "wait_minutes": wait_minutes,
            }
        )

    final_segment = segment_minutes[len(stops)] if len(segment_minutes) > len(stops) else None
    if final_segment is not None:
        total_travel_minutes += final_segment
        end_minutes = None if schedule_blocked else current_minutes + final_segment
    else:
        end_minutes = None if schedule_blocked else current_minutes

    return {
        "stops": result_stops,
        "total_travel_minutes": total_travel_minutes,
        "total_service_minutes": total_service_minutes,
        "total_wait_minutes": total_wait_minutes,
        "end_time": format_time(minutes_to_time(end_minutes)) if end_minutes is not None else "",
        "segment_details": segment_details,
        "has_conflicts": any(bool(stop.get("schedule_conflict")) for stop in result_stops),
    }

File: GAWELA-Tourenplaner-2.0/src/GAWELA.Tourenplaner2/Services/SchedulePlanner.cs
using GAWELA.Tourenplaner2.Models;

namespace GAWELA.Tourenplaner2.Services;

public static class SchedulePlanner
{
    public static ScheduleResult ComputeSchedule(List<TourStop> stops, List<int?> segmentMinutes, string startTime)
    {
        var resultStops = new List<TourStop>();
        var segmentDetails = new List<SegmentDetail>();
        var currentMinutes = TimeUtils.TimeToMinutes(startTime) ?? 8 * 60;

        var totalTravelMinutes = 0;
        var totalServiceMinutes = 0;
        var totalWaitMinutes = 0;
        var scheduleBlocked = false;

        for (var index = 0; index < stops.Count; index++)
        {
            var stop = stops[index];
            var segment = index < segmentMinutes.Count ? segmentMinutes[index] : null;
            if (segment.HasValue)
            {
                totalTravelMinutes += segment.Value;
            }

            var arrivalMinutes = (!scheduleBlocked && segment.HasValue) ? currentMinutes + segment.Value : null;
            var arrivalText = TimeUtils.FormatTime(TimeUtils.MinutesToTime(arrivalMinutes));
            var windowStart = TimeUtils.TimeToMinutes(stop.TimeWindowStart);
            var windowEnd = TimeUtils.TimeToMinutes(stop.TimeWindowEnd);

            var waitMinutes = 0;
            var conflict = false;
            var conflictText = string.Empty;
            var effectiveArrival = arrivalMinutes;

            if (effectiveArrival.HasValue && windowStart.HasValue && effectiveArrival.Value < windowStart.Value)
            {
                waitMinutes = windowStart.Value - effectiveArrival.Value;
                effectiveArrival = windowStart.Value;
                totalWaitMinutes += waitMinutes;
            }

            if (effectiveArrival.HasValue && windowEnd.HasValue && effectiveArrival.Value > windowEnd.Value)
            {
                conflict = true;
                conflictText = $"Ankunft {TimeUtils.FormatTime(TimeUtils.MinutesToTime(effectiveArrival))} > Fenster Ende {stop.TimeWindowEnd}";
            }

            var serviceMinutes = Math.Max(0, stop.ServiceMinutes);
            var departureMinutes = effectiveArrival.HasValue ? effectiveArrival.Value + serviceMinutes : null;
            if (departureMinutes.HasValue)
            {
                totalServiceMinutes += serviceMinutes;
                currentMinutes = departureMinutes.Value;
            }
            else
            {
                scheduleBlocked = true;
            }

            var updated = new TourStop
            {
                Id = stop.Id,
                Name = stop.Name,
                Address = stop.Address,
                Lat = stop.Lat,
                Lon = stop.Lon,
                Order = stop.Order,
                TimeWindowStart = stop.TimeWindowStart,
                TimeWindowEnd = stop.TimeWindowEnd,
                ServiceMinutes = stop.ServiceMinutes,
                PlannedArrival = arrivalText,
                PlannedDeparture = TimeUtils.FormatTime(TimeUtils.MinutesToTime(departureMinutes)),
                ScheduleConflict = conflict,
                ScheduleConflictText = conflictText,
                WaitMinutes = waitMinutes,
            };

            resultStops.Add(updated);
            segmentDetails.Add(new SegmentDetail
            {
                Index = index,
                TravelMinutes = segment,
                Arrival = updated.PlannedArrival,
                Departure = updated.PlannedDeparture,
                WaitMinutes = waitMinutes,
            });
        }

        var finalSegment = stops.Count < segmentMinutes.Count ? segmentMinutes[stops.Count] : null;
        int? endMinutes;
        if (finalSegment.HasValue)
        {
            totalTravelMinutes += finalSegment.Value;
            endMinutes = scheduleBlocked ? null : currentMinutes + finalSegment.Value;
        }
        else
        {
            endMinutes = scheduleBlocked ? null : currentMinutes;
        }

        return new ScheduleResult
        {
            Stops = resultStops,
            TotalTravelMinutes = totalTravelMinutes,
            TotalServiceMinutes = totalServiceMinutes,
            TotalWaitMinutes = totalWaitMinutes,
            EndTime = TimeUtils.FormatTime(TimeUtils.MinutesToTime(endMinutes)),
            SegmentDetails = segmentDetails,
            HasConflicts = resultStops.Any(s => s.ScheduleConflict),
        };
    }
}
